torchtext/csrc/vocab.h

```diff
@@ -10,29 +10,53 @@ typedef std::tuple<std::string, std::vector<int64_t>, std::vector<std::string>,
     VocabStates;
 
 struct Vocab : torch::CustomClassHolder {
-private:
+  static const int32_t MAX_VOCAB_SIZE = 30000000;
   int64_t unk_index_;
-  IndexDict stoi_;
-
-public:
+  std::vector<int32_t> stoi_;
   const std::string version_str_ = "0.0.1";
   StringList itos_;
   std::string unk_token_;
 
   explicit Vocab(const std::vector<std::string> &tokens,
                  const std::string &unk_token);
-  explicit Vocab(const StringList &tokens, const IndexDict &stoi,
-
-                 const std::string &unk_token, const int64_t unk_index);
   int64_t __len__() const;
-  int64_t __getitem__(const std::string &token) const;
+  int64_t __getitem__(const c10::string_view &token) const;
   void append_token(const std::string &token);
   void insert_token(const std::string &token, const int64_t &index);
   std::string lookup_token(const int64_t &index);
   std::vector<std::string> lookup_tokens(const std::vector<int64_t> &indices);
-  std::vector<int64_t> lookup_indices(const std::vector<std::string> &tokens);
+  std::vector<int64_t>
+  lookup_indices(const std::vector<c10::string_view> &tokens);
   std::unordered_map<std::string, int64_t> get_stoi() const;
   std::vector<std::string> get_itos() const;
+
+protected:
+  uint32_t _hash(const c10::string_view &str) const {
+    uint32_t h = 2166136261;
+    for (size_t i = 0; i < str.size(); i++) {
+      h = h ^ uint32_t(uint8_t(str[i]));
+      h = h * 16777619;
+    }
+    return h;
+  }
+
+  uint32_t _find(const c10::string_view &w) const {
+    uint32_t stoi_size = stoi_.size();
+    uint32_t id = _hash(w) % stoi_size;
+    while (stoi_[id] != -1 && c10::string_view{itos_[stoi_[id]].data(),
+                                               itos_[stoi_[id]].size()} != w) {
+      id = (id + 1) % stoi_size;
+    }
+    return id;
+  }
+
+  void _add(const std::string &w) {
+    uint32_t h = _find(c10::string_view{w.data(), w.size()});
+    if (stoi_[h] == -1) {
+      itos_.push_back(w);
```

Comment:
What happens if we reached max size?

Reply:
It's a good question, which I think can be addressed in the context of dynamic v/s static dictionary (follow-up item). As of now, the static size is 30 million, which is probably reasonable given the current vocab size typically is orders of magnitude lower.